core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/AST/ShaclProperties.java

```diff
@@ -48,39 +61,70 @@
 	private Literal maxInclusive;
 
 	private final List<String> pattern = new ArrayList<>();
-	private String flags = "";
+	private String flags = null;
 
 	private final Set<Resource> targetClass = new HashSet<>();
 	private final TreeSet<Value> targetNode = new TreeSet<>(new ValueComparator());
 	private final Set<IRI> targetSubjectsOf = new HashSet<>();
 	private final Set<IRI> targetObjectsOf = new HashSet<>();
 	private final List<Resource> targetShape = new ArrayList<>();
 
+	private Resource qualifiedValueShape;
+	private Long qualifiedMinCount;
+	private Long qualifiedMaxCount;
+	private Boolean qualifiedValueShapesDisjoint;
+
 	private final List<Resource> target = new ArrayList<>();
 
 	private boolean deactivated = false;
 
 	private boolean uniqueLang = false;
 
+	boolean closed = false;
+	private Resource ignoredProperties;
+
+	private Resource id;
+
+	private final List<Literal> message = new ArrayList<>();
+
 	public ShaclProperties() {
 	}
 
-	public ShaclProperties(Resource propertyShapeId, SailRepositoryConnection connection) {
-
-		try (Stream<Statement> stream = connection.getStatements(propertyShapeId, null, null).stream()) {
+	public ShaclProperties(Resource id, RepositoryConnection connection) {
+		this.id = id;
+		try (Stream<Statement> stream = connection.getStatements(id, null, null, true).stream()) {
 			stream.forEach(statement -> {
 				String predicate = statement.getPredicate().toString();
 				Value object = statement.getObject();
 				switch (predicate) {
+				case "http://www.w3.org/1999/02/22-rdf-syntax-ns#type":
```

Comment:
Can we replace this massive switch statement with a simple strategy pattern? Bit more coding but in the end probably a lot more readable and maintainable.

Also: can we replace the hardcoded strings with constant references?

Reply:
This code was written for performance. It was originally a bunch of getStatement calls, but those ended up being considerably slower. This class essentially just converts the raw graph in a java object to make things easier to work with. Hence the naming of the variables and getters. They all reflect the rdf property.

Reply:
I'm not convinced about the performance argument to be honest - I don't doubt that this code performs significantly better than what was previously there, but I do doubt that using hardcoded strings instead of reusing `RDF.TYPE.stringValue()`  is a significant factor in that (or indeed that a case statement is so much quicker than a simple dictionary lookup / strategy pattern impl).  However, I also appreciate that the focus should be on getting this in a mergable state, so maybe this isn't the highest priority.

I would suggest to liberally comment and javadoc the code though, to make it clear what is going on (in particular javadoc on the getters to say which RDF property they correspond to might be useful).

Reply:
It's just a java limitation. Can't use SHACL.PROPERTY.toString() because it's not bound at compile time.